controllers/baseboardmanagement_controller.go

```diff
@@ -18,42 +18,233 @@ package controllers
 
 import (
 	"context"
+	"fmt"
+	"strings"
 
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/log"
 
 	bmcv1alpha1 "github.com/tinkerbell/rufio/api/v1alpha1"
 )
 
+// BMCClient represents a baseboard management controller client.
+// It defines a set of methods to connect and interact with a BMC.
+type BMCClient interface {
+	// Close ends the connection with the bmc.
+	Close(ctx context.Context) error
+	// GetPowerState fetches the current power status of the bmc.
+	GetPowerState(ctx context.Context) (string, error)
+	// SetPowerState power controls the bmc to the input power state.
+	SetPowerState(ctx context.Context, state string) (bool, error)
+	// SetBootDevice sets the boot device on the bmc.
+	// Currently this sets the first boot device.
+	// setPersistent, if true will set the boot device permanently. If false, sets one time boot.
+	// efiBoot, if true passes efiboot options while setting boot device.
+	SetBootDevice(ctx context.Context, bootDevice string, setPersistent, efiBoot bool) (bool, error)
+}
+
+// BMCClientFactoryFunc defines a func that returns a BMCClient
+type BMCClientFactoryFunc func(ctx context.Context, hostIP, port, username, password string) (BMCClient, error)
+
 // BaseboardManagementReconciler reconciles a BaseboardManagement object
 type BaseboardManagementReconciler struct {
-	client.Client
-	Scheme *runtime.Scheme
+	client           client.Client
+	scheme           *runtime.Scheme
```

Comment:
Why do you need a scheme here? If you're passing in a client, shouldn't the scheme be registered with the client? I don't see this used anywhere

Reply:
Right it was part of the Scaffolding. I just checked it can be removed 👍